### PR TITLE
decouple product view gallery images config from Gallery Block

### PR DESCRIPTION
**Pull request by davidverholen**
Last week we had the Problem, that we wanted to add some additional image types (picture tag for different viewports) to the Product View Gallery Images.

Since the creation of the Data array for the json configuration is hardcoded in the block, we had to implement a Plugin.
https://github.com/magento/magento2/blob/develop/app/code/Magento/Catalog/Block/Product/View/Gallery.php#L58

Although it also works with a Plugin, I thought it would be nicer if this configuration is decoupled from the Gallery Block, since this configuration also depends on the template having the images defined in the view.xml

### CODE REVIEW

**Comment by davidverholen** [comment]
a better approach would maybe be to set a flag in the view xml for "use as gallery image". Not sure about this. Are there any ideas yet?

another approach would be to have the "base" image ids defined in the framework and merge it with the used theme.

I think it's not good to rely on the view.xml definitions in core modules as you can't be sure that the image definition exists in the theme

**Comment by vrann** [comment]
@davidverholen can you please merge this branch with the latest develop?

**Comment by davidverholen** [comment]
@vrann just rebased it on the current develop

File: app/code/Magento/Catalog/Block/Product/View/Gallery.php [hunk 1]
```diff
@@ -28,19 +29,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\AbstractView
...
+     * @param \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory
+     * @param array                                                               $galleryImagesConfig
+     * @param array                                                               $data
      */
     public function __construct(
         \Magento\Catalog\Block\Product\Context $context,
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
-        EncoderInterface $jsonEncoder,
+        \Magento\Framework\Json\EncoderInterface $jsonEncoder,
+        \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory,
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Block/Product/View/Gallery.php]: Unfortunately, due to backwards compatibility policy, we can not accept changes which change constructor signature by introducing a new required parameter.
You may fix it like:
```php
use Magento\Framework\App\ObjectManager;
...
public function __construct(
    \New\Dependency $newDependency = null
) {
    $this->newDependency = $newDependency ?: ObjectManager::getInstance()->get(\New\Dependency::class)
}
```

File: app/code/Magento/Catalog/Model/Product/Gallery/ImagesConfigFactory.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+use \Magento\Framework\Data\CollectionFactory;
+use Magento\Framework\DataObject;
+
+class ImagesConfigFactory implements ImagesConfigFactoryInterface
+{
+    /**
+     * @var \Magento\Framework\Data\CollectionFactory
+     */
+    protected $_dataCollectionFactory;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Model/Product/Gallery/ImagesConfigFactory.php]: We do not encourage inheritance-based API, please make this property private. Additionally, naming convention with underscores is considered legacy, please rename this property.

File: app/code/Magento/Catalog/Block/Product/View/Gallery.php [hunk 3]
```diff
@@ -28,19 +29,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\AbstractView
...
+     * @param \Magento\Framework\Json\EncoderInterface                            $jsonEncoder
+     * @param \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory
+     * @param array                                                               $galleryImagesConfig
+     * @param array                                                               $data
      */
     public function __construct(
         \Magento\Catalog\Block\Product\Context $context,
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
-        EncoderInterface $jsonEncoder,
+        \Magento\Framework\Json\EncoderInterface $jsonEncoder,
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Block/Product/View/Gallery.php]: We encourage use of `\Magento\Framework\Serialize\Serializer\Json` for json serialization.

File: app/code/Magento/ProductVideo/Block/Product/View/Gallery.php [hunk 4]
```diff
@@ -19,23 +19,29 @@ class Gallery extends \Magento\Catalog\Block\Product\View\Gallery
...
+     * @param \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory
+     * @param array                                                               $galleryImagesConfig
+     * @param array                                                               $data
      */
     public function __construct(
         \Magento\Catalog\Block\Product\Context $context,
         \Magento\Framework\Stdlib\ArrayUtils $arrayUtils,
         \Magento\Framework\Json\EncoderInterface $jsonEncoder,
         \Magento\ProductVideo\Helper\Media $mediaHelper,
+        \Magento\Catalog\Model\Product\Gallery\ImagesConfigFactoryInterface $imagesConfigFactory,
```
Comment (ishakhsuvarov) [inline app/code/Magento/ProductVideo/Block/Product/View/Gallery.php]: Please inject with the `ObjectManager` as described above.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
@davidverholen Thanks for the PR, looks pretty good to me.
Please resolve the conflict, check review comments and Travic CI fails. If it's all good I am going to proceed with the merge asap, so that no new conflicts arise.
Thank you for your collaboration!

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=3401) <br/>All committers have signed the CLA.

**Comment by davidverholen** [comment]
@ishakhsuvarov thx for the review!

I just implemented most change requests, but I had to keep the Json Encoder I think since it's part of the original constructor.

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by ishakhsuvarov** [comment]
@davidverholen Thanks, we are going to continue with review and merge

**Comment by ishakhsuvarov** [comment]
@davidverholen 
Please sign the CLA for this PR, we can not proceed until it's signed.
Also, if you don't mind, please check the "[Allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)" checkbox on this PR, so that I could push back some coding style fixes and develop merge I have made for your branch.

**Comment by davidverholen** [comment]
@ishakhsuvarov  I tried to sign the CLA, not sure if it worked. It redirected me to the CLA page, I signed in and then it reloaded the page with the box in the middle beeing empty. After some time, I was redirected back to this pr but it's still "yellow" in the pr checks

**Comment by ishakhsuvarov** [comment]
@davidverholen 
Looks like you have been working on it with someone else or with a different account. That may be the cause. It shows that user bragento has not signed.

**Comment by bragento** [comment]
@ishakhsuvarov good point, there is one commit I made with our company account

**Comment by Ctucker9233** [comment]
@magento-team Can this be backported to 2.2?

**Pull request merged**